File: Java/TwoDBinarySearch.java
package Java;

import java.util.Arrays;

/**
 * TwoDBinarySearch
 */
public class TwoDBinarySearch {
    public static void main(String[] args) {
        int[][] matrix = { { 10, 20, 30, 40 }, { 15, 25, 35, 45 }, { 28, 29, 37, 49 }, { 33, 34, 38, 50 } };
        int target = 37;
        System.out.println(matrix.length);
        System.out.println(matrix[0].length);
        int[] res = search1(matrix, target);
        System.out.println(Arrays.toString(res));

    }

    public static int[] search(int[][] matrix, int target) {
        int row = 0;
        // This is matrix lent can be different to make things
        int col = matrix[0].length - 1;
        while (row < matrix.length && col >= 0) {
            System.out.println(matrix[row][col]);
            if (matrix[row][col] < target) {
                row++;
            } else if (matrix[row][col] > target) {
                col--;
            } else {
                return new int[] { row, col };
            }
        }
        return new int[] { -1, -1 };
    }

    // Let's try to do with the things
    // Here what we did is that change the way we do binary search to other form row
    // last and first column
    public static int[] search1(int[][] matrix, int target) {
        int row = matrix[0].length - 1;
        int col = 0;
        while (col < matrix[0].length && row >= 0) {
            System.out.println(matrix[row][col]);
            if (matrix[row][col] < target) {
                col++;
            } else if (matrix[row][col] > target) {
                row--;
            } else {
                return new int[] { row, col };
            }
        }

        return new int[] { -1, -1 };
    }

}

File: Js/TwoDBinarySearch.js
const twoDSearch = (matrix, target) => {
  let row = 0
  let col = matrix[0].length - 1

  while (row < matrix.length && col >= 0) {
    console.log(matrix[row][col])
    if (matrix[row][col] > target) {
      col--
    } else if (matrix[row][col] < target) {
      row++
    } else {
      return [row, col]
    }
  }
  return [-1, -1]
}

// Question
// https://leetcode.com/problems/search-a-2d-matrix/

// let matrix = [[1]]
let matrix = [
  [1, 3, 5, 7],
  [10, 11, 16, 20],
  [23, 30, 34, 60],
]

// matrix = { { 10, 20, 30, 40 }, { 15, 25, 35, 45 }, { 28, 29, 37, 49 }, { 33, 34, 38, 50 } }
// let matrix = [
//   [10, 20, 30],
//   [15, 25, 35],
//   [28, 29, 37],
//   [33, 34, 38],
// ]
let target = 1
let x = twoDSearch(matrix, target)
console.log(x)
